File: src/test/java/flat/viewer/FlatViewServiceTest.java
package flat.viewer;

import org.junit.Before;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.mockito.Spy;
import org.mockito.junit.MockitoJUnitRunner;

import java.time.LocalDateTime;

import static flat.viewer.Result.*;
import static flat.viewer.SlotState.*;
import static org.hamcrest.CoreMatchers.equalTo;
import static org.junit.Assert.assertThat;
import static org.mockito.ArgumentMatchers.any;
import static org.mockito.Mockito.*;

@RunWith(MockitoJUnitRunner.class)
public class FlatViewServiceTest {

    @Spy
    private final NotificationService notificationService = new NotificationServiceImpl();
    private FlatViewService flatViewService;

    @Before
    public void setUp() {
//        flatViewService = new FlatViewService(notificationService, sharedData);
        flatViewService = new FlatViewService(notificationService, null);
    }

    @Test
    public void testTryReserve_notRented() {
        Integer tenant1Id = 11;
        Integer tenant2Id = 12;
        Integer flatId = 45;
        LocalDateTime start = LocalDateTime.now().plusDays(5);

        Result result = flatViewService.tryReserve(flatId, start, tenant2Id, null);

        assertThat(result, equalTo(Ok));
        verify(notificationService, never()).notifyTenant(any(), any(), any());

        flatViewService.cancel(flatId, start, tenant2Id, null);
        flatViewService.rent(flatId, tenant1Id, null);
        flatViewService.tryReserve(flatId, start, tenant2Id, null);

        verify(notificationService).notifyTenant(tenant1Id, start, RESERVING);

        result = flatViewService.approve(flatId, start, tenant2Id, null, null);

        assertThat(result, equalTo(NotCurrent));
        verify(notificationService).notifyTenant(any(), any(), any());
    }

    @Test
    public void testTryReserve_approve() {
        Integer tenant1Id = 11;
        Integer tenant2Id = 12;
        Integer tenant3Id = 13;
        Integer flatId = 45;
        LocalDateTime start = LocalDateTime.now().plusDays(5);

        Result result;
        result = flatViewService.rent(flatId, tenant1Id, null);
        assertThat(result, equalTo(Ok));
        result = flatViewService.tryReserve(flatId, start, tenant2Id, null);

        assertThat(result, equalTo(Ok));
        verify(notificationService).notifyTenant(tenant1Id, start, RESERVING);

        result = flatViewService.tryReserve(flatId, start, tenant3Id, null);
        assertThat(result, equalTo(Occupied));

        result = flatViewService.approve(flatId, start, tenant1Id, null, null);

        assertThat(result, equalTo(Ok));
        verify(notificationService).notifyTenant(tenant2Id, start, APPROVED);
    }

    @Test
    public void testTryReserve_cancel() {
        Integer tenant1Id = 11;
        Integer tenant2Id = 12;
        Integer tenant3Id = 13;
        Integer flatId = 45;
        LocalDateTime start = LocalDateTime.now().plusDays(5);

        Result result;
        result = flatViewService.rent(flatId, tenant1Id, null);
        assertThat(result, equalTo(Ok));
        flatViewService.tryReserve(flatId, start, tenant2Id, null);

        result = flatViewService.cancel(flatId, start, tenant2Id, null);

        assertThat(result, equalTo(Ok));
        verify(notificationService).notifyTenant(tenant1Id, start, CANCELED);

        result = flatViewService.tryReserve(flatId, start, tenant3Id, null);
        assertThat(result, equalTo(Ok));
    }

    @Test
    public void testTryReserve_reject() {
        Integer tenant1Id = 11;
        Integer tenant2Id = 12;
        Integer tenant3Id = 13;
        Integer flatId = 45;
        LocalDateTime start = LocalDateTime.now().plusDays(5);
        Result result;
        flatViewService.rent(flatId, tenant1Id, null);
        flatViewService.tryReserve(flatId, start, tenant2Id, null);
        verify(notificationService).notifyTenant(tenant1Id, start, RESERVING);
        result = flatViewService.reject(flatId, start, tenant1Id, null, null);

        assertThat(result, equalTo(Ok));
        verify(notificationService).notifyTenant(tenant2Id, start, REJECTED);

        result = flatViewService.tryReserve(flatId, start, tenant3Id, null);

        assertThat(result, equalTo(Rejected));
        verify(notificationService, times(2)).notifyTenant(any(), any(), any());
    }
}

File: src/test/java/flat/viewer/FlatViewServiceTest.kt
package flat.viewer

import flat.viewer.Result.*
import flat.viewer.SlotState.*
import org.hamcrest.CoreMatchers.equalTo
import org.junit.Assert.assertThat
import org.junit.Before
import org.junit.Test
import org.junit.runner.RunWith
import org.mockito.ArgumentMatchers
import org.mockito.Mockito
import org.mockito.Mockito.*
import org.mockito.Spy
import org.mockito.junit.MockitoJUnitRunner
import java.time.LocalDateTime
import java.util.*

@RunWith(MockitoJUnitRunner::class)
class FlatViewServiceTest {
    @Spy
    private val notificationService: NotificationService = NotificationServiceImpl()
    private lateinit var flatViewService: FlatViewService
    private lateinit var flatToCurrentTenant: MutableMap<Int?, Int?>
    private lateinit var flatViewToNewTenant: MutableMap<FlatViewSlot?, ViewSlot?>

    @Before
    fun setUp() {
        flatViewService = FlatViewService(notificationService)
        flatToCurrentTenant = HashMap()
        flatViewToNewTenant = HashMap()
    }

    @Test
    fun testTryReserve_notRented() {
        val tenant1Id = 11
        val tenant2Id = 12
        val flatId = 45
        val start = LocalDateTime.now().plusDays(5)

        var result = flatViewService.tryReserve(flatId, start, tenant2Id, flatViewToNewTenant)

        assertThat(result, equalTo(Ok))
        verify(notificationService, never()).notifyTenant(ArgumentMatchers.any(Int::class.java), any(), any())

        flatViewService.cancel(flatId, start, tenant2Id, flatViewToNewTenant)
        flatViewService.rent(flatId, tenant1Id, flatToCurrentTenant)
        flatViewService.tryReserve(flatId, start, tenant2Id, flatViewToNewTenant)

        verify(notificationService).notifyTenant(tenant1Id, start, RESERVING)

        result = flatViewService.approve(flatId, start, tenant2Id, flatToCurrentTenant, flatViewToNewTenant)

        assertThat(result, equalTo(NotCurrent))
        verify(notificationService).notifyTenant(ArgumentMatchers.any(Int::class.java), any(), any())
    }

    private fun <T> any(): T {
        return Mockito.any()
    }

    @Test
    fun testTryReserve_approve() {
        val tenant1Id = 11
        val tenant2Id = 12
        val tenant3Id = 13
        val flatId = 45
        val start = LocalDateTime.now().plusDays(5)

        var result: Result?
        result = flatViewService.rent(flatId, tenant1Id, flatToCurrentTenant)
        assertThat(result, equalTo(Ok))
        result = flatViewService.tryReserve(flatId, start, tenant2Id, flatViewToNewTenant)

        assertThat(result, equalTo(Ok))
        verify(notificationService).notifyTenant(tenant1Id, start, RESERVING)

        result = flatViewService.tryReserve(flatId, start, tenant3Id, flatViewToNewTenant)
        assertThat(result, equalTo(Occupied))

        result = flatViewService.approve(flatId, start, tenant1Id, flatToCurrentTenant, flatViewToNewTenant)

        assertThat(result, equalTo(Ok))
        verify(notificationService).notifyTenant(tenant2Id, start, APPROVED)
    }

    @Test
    fun testTryReserve_cancel() {
        val tenant1Id = 11
        val tenant2Id = 12
        val tenant3Id = 13
        val flatId = 45
        val start = LocalDateTime.now().plusDays(5)

        var result: Result
        result = flatViewService.rent(flatId, tenant1Id, flatToCurrentTenant)
        assertThat(result, equalTo(Ok))
        flatViewService.tryReserve(flatId, start, tenant2Id, flatViewToNewTenant)

        result = flatViewService.cancel(flatId, start, tenant2Id, flatViewToNewTenant)

        assertThat(result, equalTo(Ok))
        verify(notificationService).notifyTenant(tenant1Id, start, CANCELED)

        result = flatViewService.tryReserve(flatId, start, tenant3Id, flatViewToNewTenant)!!
        assertThat(result, equalTo(Ok))
    }

    @Test
    fun testTryReserve_reject() {
        val tenant1Id = 11
        val tenant2Id = 12
        val tenant3Id = 13
        val flatId = 45
        val start = LocalDateTime.now().plusDays(5)
        var result: Result
        flatViewService.rent(flatId, tenant1Id, flatToCurrentTenant)
        flatViewService.tryReserve(flatId, start, tenant2Id, flatViewToNewTenant)
        verify(notificationService).notifyTenant(tenant1Id, start, RESERVING)
        result = flatViewService.reject(flatId, start, tenant1Id, flatToCurrentTenant, flatViewToNewTenant)

        assertThat(result, equalTo(Ok))
        verify(notificationService).notifyTenant(tenant2Id, start, REJECTED)

        result = flatViewService.tryReserve(flatId, start, tenant3Id, flatViewToNewTenant)!!

        assertThat(result, equalTo(Rejected))
        verify(notificationService, times(2)).notifyTenant(ArgumentMatchers.any(Int::class.java), any(), any())
    }
}
